jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java

```diff
@@ -0,0 +1,277 @@
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.transport.jms.factory;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.transport.jms.exception.JMSConnectorException;
+import org.wso2.carbon.transport.jms.utils.JMSConstants;
+
+import java.util.Properties;
+
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+
+/**
+ * {@link JMSConnectionFactory} connector, session, consumer, producer caching implementation. This class can be used
+ * if needed to cache connections.
+ */
+public class CachedJMSConnectionFactory extends JMSConnectionFactory {
+    private static final Log logger = LogFactory.getLog(CachedJMSConnectionFactory.class);
+
+    private int cacheLevel = 0;
+
+    private Connection cachedConnection = null;
+    private Session cachedSession = null;
+    private MessageConsumer cachedMessageConsumer = null;
+    private MessageProducer cachedMessageProducer = null;
+
+    /**
+     * Constructor to create CachedJMSConnectionFactory instance.
+     *
+     * @param properties Property values required to initialize the connection factory
+     * @throws JMSConnectorException
+     */
+    public CachedJMSConnectionFactory(Properties properties) throws JMSConnectorException {
+        super(properties);
+        setValues(properties);
+    }
+
+    private void setValues(Properties properties) {
+        String cacheLevel = properties.getProperty(JMSConstants.PARAM_CACHE_LEVEL);
+        if (null != cacheLevel && !"".equals(cacheLevel)) {
+            this.cacheLevel = Integer.parseInt(cacheLevel);
+        } else {
+            this.cacheLevel = JMSConstants.CACHE_NONE;
+        }
+    }
+
+    @Override public ConnectionFactory getConnectionFactory() throws JMSConnectorException {
+        return super.getConnectionFactory();
+    }
+
+    @Override public Connection getConnection() throws JMSException {
+        return getConnection(null, null);
+    }
+
+    @Override public Connection getConnection(String userName, String password) throws JMSException {
+        Connection connection;
+        if (cachedConnection == null) {
+            connection = createConnection(userName, password);
+        } else {
+            connection = cachedConnection;
```

Comment:
Yes, for now, similar to the existing inbound endpoint implementation, the user must select connection caching option only if the next message to be sent, uses the same connection used to send the current message.